control/diagnostics_tool/src/system_state.c

```diff
@@ -0,0 +1,264 @@
+
+    const can_frame_s * const steering_report_frame =
+            get_can_msg_array_index_reference( PS_CTRL_MSG_ID_STEERING_REPORT );
+
+    if( steering_command_frame == NULL  ||  steering_report_frame == NULL )
+    {
+        ret = ERROR;
+    }
+
+    if( ret == NOERR )
+    {
+        ret = analyze_steering_state(
+                &system_state.steering_module_state,
+                steering_command_frame,
+                steering_report_frame );
+    }
+
+    return ret;
+}
+
+
+//
+static int update_throttle_state()
+{
+    int ret = NOERR;
+
+    const can_frame_s * const throttle_command_frame =
+            get_can_msg_array_index_reference( PS_CTRL_THROTTLE_COMMAND_ID );
+
+    const can_frame_s * const throttle_report_frame =
+            get_can_msg_array_index_reference( PS_CTRL_MSG_ID_THROTTLE_REPORT );
+
+    if( throttle_command_frame == NULL || throttle_report_frame == NULL )
+    {
+        ret = ERROR;
+    }
+
+    if( ret == NOERR )
+    {
+        ret = analyze_throttle_state(
+                &system_state.throttle_module_state,
+                throttle_command_frame,
+                throttle_report_frame );
+    }
+
+    return ret;
+}
+
+
+//
+static int update_brake_state()
+{
+    int ret = NOERR;
+
+    const can_frame_s * const brake_command_frame =
+            get_can_msg_array_index_reference( PS_CTRL_MSG_ID_BRAKE_COMMAND );
+
+    const can_frame_s * const brake_report_frame =
+            get_can_msg_array_index_reference( PS_CTRL_MSG_ID_BRAKE_REPORT );
+
+    if( brake_command_frame == NULL  ||  brake_report_frame == NULL )
+    {
+        ret = ERROR;
+    }
+
+    if( ret == NOERR )
+    {
+        ret = analyze_brake_state(
+                &system_state.brake_module_state,
+                brake_command_frame,
+                brake_report_frame );
+    }
+
+    return ret;
+}
+
+
+//
+static int update_gateway_state()
+{
+    int ret = NOERR;
+
+    const can_frame_s * const heartbeat_msg_frame =
+            get_can_msg_array_index_reference( PSVC_HEARTBEAT_MSG_BASE_ID );
+
+    const can_frame_s * const chassis_state1_frame =
+            get_can_msg_array_index_reference( PSVC_CHASSIS_STATE1_MSG_ID );
+
+    const can_frame_s * const chassis_state2_frame =
+            get_can_msg_array_index_reference( PSVC_CHASSIS_STATE2_MSG_ID );
+
+    if( heartbeat_msg_frame == NULL ||
+            chassis_state1_frame == NULL ||
+            chassis_state2_frame == NULL )
+    {
+        ret = ERROR;
+    }
+
+    if( ret == NOERR )
+    {
+        ret = analyze_gateway_state(
+                &system_state.gateway_module_state,
+                heartbeat_msg_frame,
+                chassis_state1_frame,
+                chassis_state2_frame );
+    }
+
+    return ret;
+}
+
+
+
+
+// *****************************************************
+// public definitions
+// *****************************************************
+
+
+//
+int update_system_state()
+{
+    int ret = NOERR;
+
```

Comment:
Might be simpler to do:

ret = update_steering_state();

if ( ret == ERROR )
{
...module_state = STATE_FAULT;
}